### DISCUSSION TITLE
How to install tiktoken_core on MacOs ?

### DISCUSSION THREAD

**Discussion by saifulapm** [CopilotC-Nvim/CopilotChat.nvim#300] [Q&A] [answered]
Hello, I have tried to installed tiktoken in both way like this:

```
sudo luarocks install --lua-version 5.1 tiktoken_core
luarocks install tiktoken_core
```
It's install sucessfully. But in neovim, it's showing that  tiktoken_core: missing. 

```
Installing https://luarocks.org/tiktoken_core-0.2.1-1.src.rock
tiktoken_core 0.2.1-1 is already installed in /opt/homebrew
Use --force to reinstall.
```

How we can fix it ? Am I missing anythijng ?

```
CopilotChat.nvim [dependencies] ~
- OK plenary: installed
- OK copilot: copilot.lua
- WARNING tiktoken_core: missing, optional for token counting. See README for installation instructions.
- OK treesitter[markdown]: installed
- OK treesitter[diff]: installed
```

**Comment by deathbeam** [comment]
Check the README section for it, you basically need to tell luarocks to install it to location where neovim recognizes it, you can see from where is neovim loading lua binaries via this:

`:lua print(package.cpath)`

**Comment by gptlang** [comment]
btw should we change "WARNING" to "Optional"? It isn't very clear that it's not really necessary to install it and might confuse some users.

**Comment by saifulapm** [comment]
> Check the README section for it, you basically need to tell luarocks to install it to location where neovim recognizes it, you can see from where is neovim loading lua binaries via this:
> 
> `:lua print(package.cpath)`

But how I can set path in install command ?

`sudo luarocks install --lua-version 5.1 tiktoken_core —path some path here`

like this ?

**Comment by saifulapm** [comment]
> btw should we change "WARNING" to "Optional"? It isn't very clear that it's not really necessary to install it and might confuse some users.

This is not necessary? If I install this, any extra benefits ?

**Comment by saifulapm** [comment]
Anyway, I really love your work. That was only missing feature on neovim and that’s why I was using vscode. Now I have it so I back to neovim again. 

Thank you

**Comment by deathbeam** [comment, marked as answer]
Yea its optional and you can skip it, it just makes token counting more accurate but we have fallback counting in place that somewhat works. 

> > Check the README section for it, you basically need to tell luarocks to install it to location where neovim recognizes it, you can see from where is neovim loading lua binaries via this:
> > `:lua print(package.cpath)`
> 
> But how I can set path in install command ?
> 
> `sudo luarocks install --lua-version 5.1 tiktoken_core —path some path here`
> 
> like this ?

I think it has something like that yea, check help on luarocks, it was quite annoying to set up for me as well (luarocks is rly bad)



> btw should we change "WARNING" to "Optional"? It isn't very clear that it's not really necessary to install it and might confuse some users.

I dont think neovim has OPTIONAL level, just ERROR,WARNING,OK in the healthcheck report. But it already says optional in the description so should be fine

**Comment by GitMurf** [comment]
FYI in case anyone stumbles upon this GH Discussion for MacOS when searching / troubleshooting for Windows, I have finally got luarocks / tiktoken working for Windows using PowerShell and wrote a detailed how to guide here: https://github.com/CopilotC-Nvim/CopilotChat.nvim/discussions/386#discussion-7020356

It may even help with MacOS issues 🤷‍♂️ or at least give you a couple ideas to try.

**Comment by dpnolte** [comment]
You can also use [luarocks.nvim](https://github.com/vhyrro/luarocks.nvim) plugin to install luarocks packages with lazy:
```
  {
    "vhyrro/luarocks.nvim",
    priority = 1000,
    opts = {
      rocks = { "tiktoken_core" },
    },
  },
```

**Comment by gennaro-tedesco** [comment]
> Check the README section for it, you basically need to tell luarocks to install it to location where neovim recognizes it, you can see from where is neovim loading lua binaries via this:
`:lua print(package.cpath)`

Are you sure however that this is the right location for nvim to recognise the luarocks package? That is the location where C loaders are looked in (`:help package.cpath`)
> package.cpath                                                    *package.cpath*
        The path used by `require` to search for a C loader.
        Lua initializes the C path `package.cpath` in the same way it
        initializes the Lua path `package.path`, using the environment
        variable `LUA_CPATH` (plus another default path defined in
        `luaconf.h`).

not necessarily where additional luarocks libraries are read from (I have installed `tiktoken_core` in there to no avail, still seeing the warning).
